Devirtualize GraphicsContext::drawConsumingImageBuffer()

shallawa opened the pull request:
#### 143e04df42bbcc533981678f616037385497041e
<pre>
Devirtualize GraphicsContext::drawConsumingImageBuffer()
<a href="https://bugs.webkit.org/show_bug.cgi?id=251809">https://bugs.webkit.org/show_bug.cgi?id=251809</a>
rdar://105092612

Reviewed by Tim Horton.

The method will be resolved to ImageBuffer::drawConsuming() which will draw
a NativeImage via GraphicsContext::drawNativeImage() which is virtual.

* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::drawRaisedEllipse):
(WebCore::NullGraphicsContext::drawConsumingImageBuffer): Deleted.
* Source/WebCore/platform/graphics/GraphicsContext.h:
(WebCore::GraphicsContext::drawConsumingImageBuffer):
* Source/WebCore/platform/graphics/NullGraphicsContext.h:
(WebCore::NullGraphicsContext::drawImageBuffer):
(WebCore::NullGraphicsContext::drawConsumingImageBuffer): Deleted.

Canonical link: <a href="https://commits.webkit.org/262695@main">https://commits.webkit.org/262695@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b2d794702df6db051719dbd5ec239a6677dae1ad

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2233 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2267 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2338 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/3159 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2268 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2206 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/2399 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2314 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2018 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2253 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/2011 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2009 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3016 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/1978 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1988 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1874 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2064 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2035 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3177 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2046 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1839 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1973 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1997 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/570 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2167 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/platform/graphics/NullGraphicsContext.h:
```diff
@@ -123,7 +123,6 @@ class NullGraphicsContext : public GraphicsContext {
     void drawFocusRing(const Vector<FloatRect>&, float, float, const Color&) final { }
 
     void drawImageBuffer(ImageBuffer&, const FloatRect&, const FloatRect&, const ImagePaintingOptions& = { }) final { }
-    void drawConsumingImageBuffer(RefPtr<ImageBuffer>, const FloatRect&, const FloatRect&, const ImagePaintingOptions& = { }) final;
```
weird, where's the implementation of this :)

shallawa replied:
The implementation of it is in GraphicsContext.cpp. I think it was put in a source file instead of this header because the `RefPtr<ImageBuffer>` is passed by value which requires including ImageBuffer.h.

shallawa committed:
Devirtualize GraphicsContext::drawConsumingImageBuffer()
https://bugs.webkit.org/show_bug.cgi?id=251809
rdar://105092612

Reviewed by Tim Horton.

The method will be resolved to ImageBuffer::drawConsuming() which will draw
a NativeImage via GraphicsContext::drawNativeImage() which is virtual.

* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::drawRaisedEllipse):
(WebCore::NullGraphicsContext::drawConsumingImageBuffer): Deleted.
* Source/WebCore/platform/graphics/GraphicsContext.h:
(WebCore::GraphicsContext::drawConsumingImageBuffer):
* Source/WebCore/platform/graphics/NullGraphicsContext.h:
(WebCore::NullGraphicsContext::drawImageBuffer):
(WebCore::NullGraphicsContext::drawConsumingImageBuffer): Deleted.

Canonical link: https://commits.webkit.org/262695@main
